apps/wakunode2/wakunode2.nim

```diff
@@ -214,6 +214,51 @@ proc retrieveDynamicBootstrapNodes*(dnsDiscovery: bool, dnsDiscoveryUrl: string,
   debug "No method for retrieving dynamic bootstrap nodes specified."
   ok(newSeq[RemotePeerInfo]()) # Return an empty seq by default
 
+proc setupNat(natConf, clientId: string, tcpPort, udpPort: Port):
+  SetupResult[tuple[ip: Option[ValidIpAddress], tcpPort: Option[Port], udpPort: Option[Port]]] {.gcsafe.} =
+
+  let strategy = case natConf.toLowerAscii():
+      of "any": NatAny
+      of "none": NatNone
+      of "upnp": NatUpnp
+      of "pmp": NatPmp
+      else: NatNone
+
+  var endpoint: tuple[ip: Option[ValidIpAddress], tcpPort: Option[Port], udpPort: Option[Port]]
+
+  if strategy != NatNone:
+    let extIp = getExternalIP(strategy)
+    if extIP.isSome():
+      endpoint.ip = some(ValidIpAddress.init(extIp.get()))
+      # RedirectPorts in considered a gcsafety violation
+      # because it obtains the address of a non-gcsafe proc?
+      var extPorts: Option[(Port, Port)]
+      try:
+        extPorts = ({.gcsafe.}: redirectPorts(tcpPort = tcpPort,
+                                              udpPort = udpPort,
+                                              description = clientId))
+      except CatchableError:
+        # TODO: nat.nim Error: can raise an unlisted exception: Exception. Isolate here for now.
+        error "unable to determine external ports"
+        extPorts = none((Port, Port))
```

Comment:
Should it return `error(...)` in this case ? I'm afraid it'd return `ok(...)`

Reply:
It sets the `extPorts` to none, a valid state. So no need to break the execution here.